gsw/geostrophy.py

```diff
@@ -139,48 +141,43 @@ def distance(lon, lat, p=0):
         distance in meters between adjacent points.
 
     """
-    # This uses the algorithm from pycurrents rather than the one
-    # in GSW-Matlab.
+    lon, lat, = np.atleast_2d(lon), np.atleast_2d(lat)
 
-    if lon.ndim != 1 or lat.ndim != 1:
-        raise ValueError('lon, lat must be 1-D; found shapes %s and %s'
-                         % (lon.shape, lat.shape))
-    if lon.shape != lat.shape:
-        raise ValueError('lon, lat must have same 1-D shape; found %s and %s'
-                         % (lon.shape, lat.shape))
-    if p != 0:
-        if np.iterable(p) and p.shape != lon.shape:
-            raise ValueError('lon, non-scalar p must have same 1-D shape;'
-                             ' found %s and %s'
-                             % (lon.shape, lat.shape))
+    if (lon.size == 1) & (lat.size == 1):
+        raise ValueError('more than one point is needed to compute distance')
+    elif lon.ndim != lat.ndim:
+        raise ValueError('lon, lat must have the same dimension')
 
-        p = np.broadcast_to(p, lon.shape)
+    lon, lat, p = np.broadcast_arrays(lon, lat, p, subok=True)
 
-    slm = slice(None, -1)
-    slp = slice(1, None)
+    p_mid = 0.5 * (p[:, 0:-1] + p[:, 0:-1])
+    lat_mid = 0.5 * (lat[:, :-1] + lat[:, 1:])
 
-    radius = 6371e3
+    z = z_from_p(p_mid, lat_mid)
 
     lon = np.radians(lon)
     lat = np.radians(lat)
-    if p != 0:
-        p_mid = 0.5 * (p[slp] + p[slm])
-        lat_mid = 0.5 * (lat[slp] + lat[slm])
-        z_mid = gsw.z_from_p(p_mid, lat_mid)
-        radius += z_mid
 
-    d = np.arccos(cos(lat[slm]) * cos(lat[slp]) * cos(lon[slp] - lon[slm])
-                  + sin(lat[slm]) * sin(lat[slp])) * radius
+    dlon = np.diff(lon)
+    dlat = np.diff(lat)
+
+    a = ((np.sin(dlat / 2)) ** 2 + np.cos(lat[:, :-1]) *
+         np.cos(lat[:, 1:]) * (np.sin(dlon / 2)) ** 2)
 
-    return d
+    angles = 2 * np.arctan2(np.sqrt(a), np.sqrt(1 - a))
+
+    earth_radius = 6371e3
+    distance = (earth_radius + z) * angles
```

Comment:
It still makes sense to avoid all of the p-processing in the p==0 case.  It is a bit silly to include it at all, but we are stuck with it.  I can fix this later, though.